### PR TITLE
navigation to study/series/instance 

### PR DESCRIPTION
**Pull request by pavertomato**
[#23] navigation to study/series/instance even if it's not present in folder

### CODE REVIEW

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java [hunk 1]
```diff
@@ -80,13 +83,22 @@
     }
 
     if (studyList.size() > MAX_STUDIES_IN_DICOM_STORE) {
-      throw new ToManyResultsException("Too large DICOM Store. In " + dicomPath.getDicomStoreId() +
+      LOGGER.warn("Too large DICOM Store. In " + dicomPath.getDicomStoreId() +
```
Comment (pkling) [inline src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: Why are we making this change? It seems unrelated to the purpose of this CL, no?

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java [hunk 2]
```diff
@@ -146,12 +170,22 @@
...
     }
     if (instancesList.size() > MAX_INSTANCES_IN_SERIES) {
-      throw new ToManyResultsException("Too large Series. In " + queryBuilder.getSeriesId() +
+      LOGGER.warn("Too large Series. In " + queryBuilder.getSeriesId() +
           " over " + MAX_INSTANCES_IN_SERIES + " Instances!");
     }
     return instancesList;
   }
 
+  public static Instance getInstance(FuseDao fuseDao, CloudConf cloudConf, DicomPath dicomPath) throws DicomFuseException {
```
Comment (pkling) [inline src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: getSingleInstance for consistency with getSingleSeries?

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java [hunk 3]
```diff
@@ -80,13 +83,22 @@
...
     if (studyList.size() > MAX_STUDIES_IN_DICOM_STORE) {
-      throw new ToManyResultsException("Too large DICOM Store. In " + dicomPath.getDicomStoreId() +
+      LOGGER.warn("Too large DICOM Store. In " + dicomPath.getDicomStoreId() +
           " over " + MAX_STUDIES_IN_DICOM_STORE + " Studies!");
     }
 
     return studyList;
   }
 
+  public static Study getStudy(FuseDao fuseDao, CloudConf cloudConf,
```
Comment (pkling) [inline src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: getSingleStudy() for consistency with getSingleSeries()?

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java [hunk 4]
```diff
@@ -100,7 +103,8 @@ public FuseDaoImpl(AuthAdc authAdc, HttpClientFactory httpClientFactory) {
         .setHost(HEALTHCARE_HOST)
         .setPath(dicomStoresPathBuilder.toPath());
     DicomStores dicomStoresOnPage = createRequestForObjectList(uriBuilder,
-        new TypeReference<DicomStores>() {});
+        new TypeReference<DicomStores>() {
```
Comment (pkling) [inline src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java]: Noop change, revert?

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java [hunk 5]
```diff
@@ -109,7 +113,8 @@ public FuseDaoImpl(AuthAdc authAdc, HttpClientFactory httpClientFactory) {
     while (nextPageToken != null) {
       uriBuilder.setParameter(PARAM_PAGE_TOKEN, nextPageToken);
       dicomStoresOnPage = createRequestForObjectList(uriBuilder,
-          new TypeReference<DicomStores>() {});
```
Comment (pkling) [inline src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java]: Noop change, revert?

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java [hunk 6]
```diff
@@ -126,7 +131,25 @@ public FuseDaoImpl(AuthAdc authAdc, HttpClientFactory httpClientFactory) {
         .addParameter(PARAM_LIMIT, VALUE_PARAM_MAX_LIMIT_FOR_STUDY.toString())
         .addParameter(PARAM_OFFSET, queryBuilder.getOffset().toString())
         .setPath(path);
-    return createRequestForObjectList(uriBuilder, new TypeReference<List<Study>>() {});
```
Comment (pkling) [inline src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java]: Noop change, revert?

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java [hunk 7]
```diff
@@ -140,7 +163,27 @@ public FuseDaoImpl(AuthAdc authAdc, HttpClientFactory httpClientFactory) {
         .addParameter(PARAM_LIMIT, VALUE_PARAM_MAX_LIMIT_FOR_SERIES.toString())
         .addParameter(PARAM_OFFSET, queryBuilder.getOffset().toString())
         .setPath(path);
-    return createRequestForObjectList(uriBuilder, new TypeReference<List<Series>>() {});
+    return createRequestForObjectList(uriBuilder, new TypeReference<List<Series>>() {
```
Comment (pkling) [inline src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java]: Noop change, revert?

If you want to change the style across the entire file, please do that in a separate clean-up change (to keep diffs small and readable).

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java [hunk 8]
```diff
@@ -155,7 +198,27 @@ public FuseDaoImpl(AuthAdc authAdc, HttpClientFactory httpClientFactory) {
         .addParameter(PARAM_LIMIT, VALUE_PARAM_MAX_LIMIT_FOR_INSTANCES.toString())
         .addParameter(PARAM_OFFSET, queryBuilder.getOffset().toString())
         .setPath(path);
-    return createRequestForObjectList(uriBuilder, new TypeReference<List<Instance>>() {});
+    return createRequestForObjectList(uriBuilder, new TypeReference<List<Instance>>() {
```
Comment (pkling) [inline src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java]: Noop change, revert?

File: src/main/java/com/google/dicomwebfuse/fuse/DicomFuseHelper.java [hunk 9]
```diff
@@ -510,6 +510,17 @@ private void updateStudiesInDicomStore(DicomPath dicomPath, boolean forceUpdate)
     cache.setDicomStoreCacheTime(dicomPath, newInstant);
   }
 
+  private void updateStudyInDicomStore(DicomPath dicomPath, boolean forceUpdate)
```
Comment (pkling) [inline src/main/java/com/google/dicomwebfuse/fuse/DicomFuseHelper.java]: ...Single... for consistency

File: src/main/java/com/google/dicomwebfuse/fuse/DicomFuseHelper.java [hunk 10]
```diff
@@ -537,6 +548,27 @@ private void updateSeriesInStudy(DicomPath dicomPath) throws DicomFuseException
...
+    Series series = FuseDaoHelper.getSingleSeries(parameters.getFuseDAO(),
+        parameters.getCloudConf(), dicomPath);
+    String seriesInstanceUID = series.getSeriesInstanceUID().getValue1();
+    CachedSeries newCachedSeries = new CachedSeries(series);
+    cache.getCachedSeries(dicomPath).put(seriesInstanceUID, newCachedSeries);
+    Instant newInstant = Instant.now().plusSeconds(parameters.getCacheTime().getObjectsCacheTime());
+    cache.setStudyCacheTime(dicomPath, newInstant);
+  }
+
+  private void updateInstanceInSeries(DicomPath dicomPath) throws DicomFuseException {
```
Comment (pkling) [inline src/main/java/com/google/dicomwebfuse/fuse/DicomFuseHelper.java]: ...Single... for consistency

File: src/test/java/com/google/dicomwebfuse/TestUtils.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+      basicHttpEntities.add(httpEntity);
+    } else if (entityCount == entityLimit) {
+      BasicHttpEntity httpEntity = prepareHttpEntityWithContent(entityCount, entityContent);
+      basicHttpEntities.add(httpEntity);
+      for (int i = 0; i < iterationCount; i++) {
+        BasicHttpEntity emptyHttpEntity = prepareEmptyHttpEntity();
+        basicHttpEntities.add(emptyHttpEntity);
+      }
+    } else {
+      for (int i = 0, count = entityCount; i < iterationCount + 1; i++, count -= entityLimit) {
```
Comment (pkling) [inline src/test/java/com/google/dicomwebfuse/TestUtils.java]: I find this control flow a bit hard to follow due to the deep nesting. Please consider extracting a utility function.

File: src/test/java/com/google/dicomwebfuse/dao/FuseDaoHelperTest.java [hunk 12]
```diff
@@ -111,20 +99,6 @@ void testShouldReturnFiveThousandStudiesIfStudiesCountFiveThousand()
     assertEquals(expectedStudiesCount, actualStudyList.size());
   }
 
-  @Test
```
Comment (pkling) [inline src/test/java/com/google/dicomwebfuse/dao/FuseDaoHelperTest.java]: Why remove these test cases?

File: src/test/java/com/google/dicomwebfuse/dao/FuseDaoHelperTest.java [hunk 13]
```diff
@@ -111,20 +99,6 @@ void testShouldReturnFiveThousandStudiesIfStudiesCountFiveThousand()
     assertEquals(expectedStudiesCount, actualStudyList.size());
   }
```
Comment (pkling) [inline src/test/java/com/google/dicomwebfuse/dao/FuseDaoHelperTest.java]: Please add new test cases for navigating to a study that's not present in the listing.

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java (hunk 1 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: logic of throwing exception in case if folder have >15k objects breaks workflow because when we open /some/path/to/study OS checks if all folders from this path is exists, so we have to return code 0 for folder with >15k objects. we can't make /some/path/to/study available and /some/path unavailable in same time
it may be done how it works now - by showing downloaded result or we may show empty folder. not sure which way is less confusing for user.

File: src/test/java/com/google/dicomwebfuse/dao/FuseDaoHelperTest.java (hunk 13 above)
Comment (pavertomato) [reply src/test/java/com/google/dicomwebfuse/dao/FuseDaoHelperTest.java]: it's done in src/test/java/com/google/dicomwebfuse/fuse/DicomFuseHelperTest.java  https://github.com/GoogleCloudPlatform/healthcare-api-dicom-fuse/pull/28/files/463d95ee32b65c83d4a621575be8e7f72278e855#diff-fb5541c3c9fa2fdbb8cdf87c64d2ce54R1 because changes are made in DicomFuseHelper class

File: src/test/java/com/google/dicomwebfuse/dao/FuseDaoHelperTest.java (hunk 12 above)
Comment (pavertomato) [reply src/test/java/com/google/dicomwebfuse/dao/FuseDaoHelperTest.java]: as i responded https://github.com/GoogleCloudPlatform/healthcare-api-dicom-fuse/pull/28/files/463d95ee32b65c83d4a621575be8e7f72278e855#r326368130 if we want to open unlisted files we no longer need logic with exception for large folder

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java (hunk 6 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java]: done

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java (hunk 2 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: done

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java (hunk 3 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: done

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java (hunk 4 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java]: reverted

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java (hunk 5 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java]: reverted

File: src/main/java/com/google/dicomwebfuse/fuse/DicomFuseHelper.java (hunk 9 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/fuse/DicomFuseHelper.java]: done

File: src/main/java/com/google/dicomwebfuse/fuse/DicomFuseHelper.java (hunk 10 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/fuse/DicomFuseHelper.java]: done

File: src/test/java/com/google/dicomwebfuse/TestUtils.java (hunk 11 above)
Comment (pavertomato) [reply src/test/java/com/google/dicomwebfuse/TestUtils.java]: this code from https://github.com/GoogleCloudPlatform/healthcare-api-dicom-fuse/pull/25 moved to separated file to make it reusable, do you prefer to fix it in this PR or better to create new one?

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java (hunk 7 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java]: reverted

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java (hunk 8 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoImpl.java]: reverted

File: src/test/java/com/google/dicomwebfuse/dao/FuseDaoHelperTest.java (hunk 13 above)
Comment (pavertomato) [reply src/test/java/com/google/dicomwebfuse/dao/FuseDaoHelperTest.java]: UPD: https://github.com/GoogleCloudPlatform/healthcare-api-dicom-fuse/pull/28/files#diff-fb5541c3c9fa2fdbb8cdf87c64d2ce54R1

**Comment by pavertomato** [comment]
@pkling could you please review updates for this PR?

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java (hunk 1 above)
Comment (dzelemba) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: Makes sense. Could we just update the error message while we're here:

"DICOM Store %s has more than 15k studies in it, so only showing the first 15k"

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java [hunk 14]
```diff
@@ -121,13 +133,25 @@
     }
 
     if (seriesList.size() > MAX_SERIES_IN_STUDY) {
-      throw new ToManyResultsException("Too large Study. In " + dicomPath.getStudyInstanceUID() +
+      LOGGER.warn("Too large Study. In " + dicomPath.getStudyInstanceUID() +
           " over " + MAX_SERIES_IN_STUDY + " Series!");
```
Comment (dzelemba) [inline src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: Update this error message too.

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java [hunk 15]
```diff
@@ -146,12 +170,22 @@
       instancesList.addAll(resultInstancesList);
     }
     if (instancesList.size() > MAX_INSTANCES_IN_SERIES) {
-      throw new ToManyResultsException("Too large Series. In " + queryBuilder.getSeriesId() +
+      LOGGER.warn("Too large Series. In " + queryBuilder.getSeriesId() +
```
Comment (dzelemba) [inline src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: And here.

File: src/test/java/com/google/dicomwebfuse/TestUtils.java (hunk 11 above)
Comment (dzelemba) [reply src/test/java/com/google/dicomwebfuse/TestUtils.java]: Let's just fix here.

File: src/test/java/com/google/dicomwebfuse/TestUtils.java [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.message.BasicStatusLine;
+import org.mockito.ArgumentMatchers;
+import org.mockito.Mockito;
+
+public class TestUtils {
+
+  static public HttpClientFactory prepareHttpClient(int entityCount, int maxEntities,
+      int entityLimit,
+      String entityContent, int statusCode) throws IOException {
```
Comment (dzelemba) [inline src/test/java/com/google/dicomwebfuse/TestUtils.java]: Weird line break here did you run the Java formatter on this?

File: src/test/java/com/google/dicomwebfuse/TestUtils.java [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.entity.BasicHttpEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.message.BasicStatusLine;
+import org.mockito.ArgumentMatchers;
+import org.mockito.Mockito;
+
+public class TestUtils {
+
+  static public HttpClientFactory prepareHttpClient(int entityCount, int maxEntities,
```
Comment (dzelemba) [inline src/test/java/com/google/dicomwebfuse/TestUtils.java]: Add Javadoc.

**Comment by pavertomato** [comment]
> Could we also add an integration test for this?

for this test we need to have store with >15k studies in gcp-healthcare-oss-test, seems we don't have permissions to create new stores, could you please create it or grant access? is it ok to use our test data, or better to use yours?
also - can we be sure what  study with index 15001 after some time won't became study with index [0;15000]? because we can't specify any type of sort in requests

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java (hunk 1 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: done

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java (hunk 14 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: done

File: src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java (hunk 15 above)
Comment (pavertomato) [reply src/main/java/com/google/dicomwebfuse/dao/FuseDaoHelper.java]: done

File: src/test/java/com/google/dicomwebfuse/TestUtils.java (hunk 16 above)
Comment (pavertomato) [reply src/test/java/com/google/dicomwebfuse/TestUtils.java]: yeah, seems like it was something wrong with idea's formatter in this case

File: src/test/java/com/google/dicomwebfuse/TestUtils.java (hunk 17 above)
Comment (pavertomato) [reply src/test/java/com/google/dicomwebfuse/TestUtils.java]: done

**Comment by dzelemba** [comment]
Ok I created DICOMStore with ID "retrieve-study-past-max-studies-limit" in the dicomfuse dataset and gave you permission to write to it.

You can use your own testdata. For the sort problem, we do guarantee that the order will be consistent (otherwise using limit/offset to page through a DICOMstore might not work). To be extra safe though (in case studies get deleted) you can do a QIDO query to get a study outside the first 15k.

File: src/test/java/com/google/dicomwebfuse/TestUtils.java (hunk 11 above)
Comment (pavertomato) [reply src/test/java/com/google/dicomwebfuse/TestUtils.java]: done

**Comment by pavertomato** [comment]
> Ok I created DICOMStore with ID "retrieve-study-past-max-studies-limit" in the dicomfuse dataset and gave you permission to write to it.
> 
> You can use your own testdata. For the sort problem, we do guarantee that the order will be consistent (otherwise using limit/offset to page through a DICOMstore might not work). To be extra safe though (in case studies get deleted) you can do a QIDO query to get a study outside the first 15k.

done
